apps/prairielearn/python/traverse.py

```diff
@@ -0,0 +1,183 @@
+                # that produce SVG documents.
+                #
+                # We transform these into comments to match the behavior of the HTML spec,
+                # as well as the `parse5` npm package that we use for parsing in JavaScript.
+                tail = new_elements.tail
+                new_elements.tail = None
+                instruction = lxml.html.tostring(new_elements, encoding="unicode")
+                instruction = instruction.removeprefix("<?").removesuffix("?>")
+                result.append("<!--?" + instruction + "?-->")
+                if tail:
+                    result.append(tail)
+            else:
+                # Add opening tag and text
+                result.append(get_source_definition(new_elements))
+
+                if new_elements.text is not None:
+                    result.append(html_escape(new_elements.text))
+
+                # Add all children to the work stack
+                children = list(new_elements)
+
+                if children:
+                    work_stack.extend(reversed(children))
+
+                count_stack.append(len(children))
+                tail_stack.append((new_elements.tag, element.tail))
+
+        # Close all closable tags
+        while count_stack[-1] == 0:
+            count_stack.pop()
+            tail_tag, tail_text = tail_stack.pop()
+
+            if tail_tag not in VOID_ELEMENTS and tail_tag is not None:
+                result.append(f"</{tail_tag}>")
+            if tail_text is not None:
+                result.append(tail_text)
+        else:
+            count_stack[-1] -= 1
+
+    # No need to empty tail stack, should be empty from above.
+    # If debugging, you can add the following assertions:
+    #
+    # assert count_stack == deque([0])
+    # assert not tail_stack
+
+    return "".join(result)
+
+
+if __name__ == "__main__":
+    html = """
+    <pl-question-panel>
+    <p> Consider two numbers $a = {{params.a}}$ and $b = {{params.b}}$.</p>
+    <p> What is the sum $c = a + b$?</p>
+    </pl-question-panel>
+
+    <pl-integer-input answers-name="c" label="$c=$"></pl-integer-input>
+
+    <pl-answer-panel>
+      <pl-code-tabs filename="foo.py"></pl-code-tabs>
+    </pl-answer-panel>
+    """
+
+    def replace(e: lxml.html.Element) -> str:
+        if e.tag == "<pl-integer-input>":
+            return "<pl-number-input></pl-number-input>"
+        if e.tag == "pl-number-input":
+            return '<input type="number" />'
+        if e.tag == "pl-code-tabs":
+            return "<div><pl-code></pl-code><pl-code></pl-code></div>"
+        if e.tag == "pl-code":
+            return "<pre><code>foo</code></pre>"
+        return e
+
+    number = 1000
+    total_time = timeit.timeit(
+        lambda: traverse_and_replace(html, replace), number=number
+    )
+    time_per_iter = total_time / number
+
+    print(f"${number} loops, {time_per_iter * 1000}ms per loop")
```

Comment:
TODO: can probably remove this?

Reply:
I think this can be removed. More robust benchmarking code for this can be added back later if needed.

Reply:
Removed in c43d770.